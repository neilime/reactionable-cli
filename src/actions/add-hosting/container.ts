import 'reflect-metadata';

import { Container } from 'inversify';

import Amplify from './adapters/amplify/Amplify';
import { HostingAdapter } from './adapters/HostingAdapter';
import Netlify from './adapters/netlify/Netlify';

export const AdapterKey = 'HostingAdapter';

export function bindHostingAdapters(container: Container): void {
  container.bind<HostingAdapter>(AdapterKey).to(Amplify);
  container.bind<HostingAdapter>(AdapterKey).to(Netlify);
}
