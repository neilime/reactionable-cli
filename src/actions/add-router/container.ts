import "reflect-metadata";

import { Container } from "inversify";

import NextJs from "./adapters/NextJs";
import { RouterAdapter } from "./adapters/RouterAdapter";
import RouterDom from "./adapters/RouterDom";

export const AdapterKey = "RouterAdapter";

export function bindRouterAdapters(container: Container): void {
  container.bind<RouterAdapter>(AdapterKey).to(NextJs);
  container.bind<RouterAdapter>(AdapterKey).to(RouterDom);
}
