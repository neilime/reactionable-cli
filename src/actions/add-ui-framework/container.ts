import 'reflect-metadata';

import { Container } from 'inversify';

import { IUIFrameworkAdapter } from './adapters/IUIFrameworkAdapter';
import ReactBootstrap from './adapters/ReactBootstrap';

export const AdapterKey = 'UIFrameworkAdapter';

export function bindUIFrameworkAdapters(container: Container) {
  container.bind<IUIFrameworkAdapter>(AdapterKey).to(ReactBootstrap);
}
