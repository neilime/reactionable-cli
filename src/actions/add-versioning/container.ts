import 'reflect-metadata';
import { Container } from 'inversify';
import Git from './adapters/github/Github';
import { IVersioningAdapter } from './IVersioningAdapter';

let container = new Container({ defaultScope: 'Singleton' });

container.bind<IVersioningAdapter>('Adapter').to(Git);

export default container;