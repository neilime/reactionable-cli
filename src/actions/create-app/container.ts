import 'reflect-metadata';

import { Container } from 'inversify';

import CreateReactApp from './adapters/create-react-app/CreateReactApp';
import { CreateAppAdapter } from './adapters/CreateAppAdapter';

export const AdapterKey = 'CreateAppAdapter';

export function bindCreateAppAdapters(container: Container): void {
  container.bind<CreateAppAdapter>(AdapterKey).to(CreateReactApp);
  container.bind<CreateReactApp>(CreateReactApp).toSelf();
}
