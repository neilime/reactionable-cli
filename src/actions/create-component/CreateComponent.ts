import { join, resolve, dirname, basename } from 'path';
import { injectable, inject } from 'inversify';
import { prompt } from 'inquirer';
import { mkdirSync } from 'fs';

import { ConsoleService } from '../../services/ConsoleService';
import { FileService } from '../../services/file/FileService';
import { PackageManagerService } from '../../services/package-manager/PackageManagerService';
import { TemplateService } from '../../services/TemplateService';
import { IAction } from '../IAction';
import AddUIFramework from '../add-ui-framework/AddUIFramework';
import AddHosting from '../add-hosting/AddHosting';
import { AbstractAdapterWithPackage } from '../AbstractAdapterWithPackage';
import { StringUtils } from '../../services/StringUtils';

@injectable()
export default class CreateComponent
  implements IAction<{ name: string | undefined }> {
  protected static defaultPackage = '@reactionable/core';
  protected static viewsPath = join('', 'src', 'views');
  protected static templateNamespace = 'create-component';

  constructor(
    @inject(AddUIFramework) private readonly addUIFramework: AddUIFramework,
    @inject(AddHosting) private readonly addHosting: AddHosting,
    @inject(PackageManagerService)
    protected readonly packageManagerService: PackageManagerService,
    @inject(ConsoleService) protected readonly consoleService: ConsoleService,
    @inject(FileService) protected readonly fileService: FileService,
    @inject(TemplateService) protected readonly templateService: TemplateService
  ) {}

  getName() {
    return 'Create a new react component';
  }

  async run({ realpath, name }) {
    if (!name) {
      const answer = await prompt<{ name: string }>([
        {
          name: 'name',
          message: "What's the component name?",
          validate: (input) =>
            input.trim().length ? true : 'Component name is required',
          transformer: (input) => this.formatName(input),
        },
      ]);
      name = answer.name;
    }

    name = this.formatName(name);
    this.consoleService.info(`Create component "${name}"...`);

    const componentDirPath = await this.createComponent({ realpath, name });

    this.consoleService.success(
      `Component "${name}" has been created in "${componentDirPath}"`
    );
  }

  protected async createComponent({
    realpath,
    componentDirPath,
    name,
    componentTemplate = 'simple/Simple.tsx',
    testComponentTemplate = 'simple/Simple.test.tsx',
    templateContext = {},
  }: {
    realpath: string;
    name: string;
    componentDirPath?: string;
    componentTemplate?: string;
    testComponentTemplate?: string;
    templateContext?: Object;
  }): Promise<string> {
    // Define component path
    const componentDirName = StringUtils.hyphenize(name);

    if (!componentDirPath) {
      componentDirPath = realpath;
    }

    if (componentDirPath.indexOf(CreateComponent.viewsPath) === -1) {
      const viewsRealpath = resolve(
        componentDirPath,
        CreateComponent.viewsPath
      );
      if (!this.fileService.dirExistsSync(viewsRealpath)) {
        mkdirSync(viewsRealpath);
      }
      componentDirPath = resolve(viewsRealpath, componentDirName);
    } else {
      componentDirPath = resolve(componentDirPath, componentDirName);
    }

    switch (name) {
      case 'App':
        componentDirPath = resolve(realpath, 'src');
        componentTemplate = 'app/App.tsx';
        break;
      case 'NotFound':
        componentTemplate = 'not-found/NotFound.tsx';
        break;
    }

    // Get enabled UI framework
    const context = {
      ...templateContext,
      componentName: name,
      projectName: await this.packageManagerService.getPackageName(
        realpath,
        'capitalizeWords'
      ),
      uiPackage: await this.getUIPackage(realpath),
      hostingPackage: await this.getHostingPackage(realpath),
    };

    // Create component from template
    await this.templateService.renderTemplateTree(
      dirname(componentDirPath),
      CreateComponent.templateNamespace,
      {
        [basename(componentDirPath)]: {
          [name + '.tsx']: componentTemplate,
          [name + '.test.tsx']: testComponentTemplate,
        },
      },
      context
    );
    return componentDirPath;
  }

  protected formatName(name: string): string {
    return StringUtils.capitalize(StringUtils.camelize(name.trim()));
  }

  protected async getUIPackage(realpath: string): Promise<string> {
    const uiPackage = (
      await this.addUIFramework.detectAdapter(this.getProjectRootPath(realpath))
    )?.getAdapterPackageName();
    if (uiPackage) {
      return uiPackage;
    }
    return CreateComponent.defaultPackage;
  }

  private async getHostingPackage(realpath: string): Promise<string> {
    const hostingAdapter = await this.addHosting.detectAdapter(
      this.getProjectRootPath(realpath)
    );

    if (
      hostingAdapter &&
      hostingAdapter instanceof AbstractAdapterWithPackage
    ) {
      return hostingAdapter.getAdapterPackageName();
    }

    return CreateComponent.defaultPackage;
  }

  private getProjectRootPath(realpath: string): string {
    let projectRootPath: string;
    if (realpath.indexOf(CreateComponent.viewsPath) === -1) {
      projectRootPath = realpath;
    } else {
      projectRootPath = realpath.split(CreateComponent.viewsPath)[0];
    }
    return projectRootPath;
  }
}
