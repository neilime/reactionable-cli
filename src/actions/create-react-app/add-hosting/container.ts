import 'reflect-metadata';
import { Container } from 'inversify';
import { IHostingAction } from './hostings/IHostingAction';
import Amplify from './hostings/amplify/Amplify';

let container = new Container({ defaultScope: 'Singleton' });

container.bind<IHostingAction>('Hosting').to(Amplify);

export default container;