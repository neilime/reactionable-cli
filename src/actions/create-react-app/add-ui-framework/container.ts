import 'reflect-metadata';
import { Container } from 'inversify';
import ReactBootstrap from './ui-frameworks/ReactBootstrap';
import { IUIFrameworkAction } from './ui-frameworks/IUIFrameworkAction';

let container = new Container({ defaultScope: 'Singleton' });

container.bind<IUIFrameworkAction>('UIFramework').to(ReactBootstrap);

export default container;