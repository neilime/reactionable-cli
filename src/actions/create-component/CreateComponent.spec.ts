import CreateComponent from './CreateComponent';
import container from '../../container';
import { resolve } from 'path';
import { promisify } from 'util';
import { existsSync, readFileSync } from 'fs';

describe('CreateComponent', () => {
  let createComponent: CreateComponent;

  const testDirPath = '__tests__/test-project';

  const cleanTesttDir = async () => {
    const testComponentDirPath = resolve(
      testDirPath,
      'src/views/test-component'
    );
    const rimraf = promisify(require('rimraf'));
    await rimraf(testComponentDirPath);
  };

  beforeAll(() => {
    createComponent = container.get(CreateComponent);
  });

  beforeEach(cleanTesttDir);
  afterAll(cleanTesttDir);

  describe('construct', () => {
    it('should be initialized ', () => {
      expect(createComponent).toBeInstanceOf(CreateComponent);
    });
  });

  describe('run', () => {
    it('should create expected components files', async () => {
      await createComponent.run({
        realpath: testDirPath,
        name: 'test component',
      });

      const expectedFiles = [
        'views/test-component/TestComponent.test.tsx',
        'views/test-component/TestComponent.tsx',
      ];

      for (const expectedFile of expectedFiles) {
        const filePath = resolve(testDirPath, 'src', expectedFile);
        expect(existsSync(filePath)).toBe(true);
        expect(readFileSync(filePath, 'utf-8')).toMatchSnapshot();
      }
    });
  });
});
