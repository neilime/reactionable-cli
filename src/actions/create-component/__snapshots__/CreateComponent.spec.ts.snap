// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createComponent run should create expected App component files 1`] = `
"import { IIdentityContextProviderProps, useIdentityContextProviderProps } from '@reactionable/amplify';
import { App as CoreApp, IAppProps } from '@reactionable/core';
import { IUIContextProviderProps, IUseLayoutProps, useUIContextProviderProps } from '@reactionable/ui-bootstrap';
import { FC, lazy } from 'react';
import { useTranslation } from 'react-i18next';
import './App.scss';
import './i18n/i18n';

// Lazy load these components
const HomeComponent = lazy(() => import('./views/home/Home'));

const NotFoundComponent = lazy(() => import('./views/not-found/NotFound'));


const App: FC = () => {
  const { t } = useTranslation();

  // Configure app
  const appConfig: IAppProps<
    IIdentityContextProviderProps,
    IUIContextProviderProps,
    IUseLayoutProps
  > = {
    routes: [],
    HomeComponent,
    NotFoundComponent,
    layout: {
      header: {
        brand: t('Tests'),
      },
      footer: {
        brand: t('Tests'),
      },
    },
    ui: useUIContextProviderProps(),
    identity: useIdentityContextProviderProps(),
  };

  return <CoreApp {...appConfig} />;
};

export default App;"
`;

exports[`createComponent run should create expected App component files 2`] = `
"import { IIdentityContextProviderProps, useIdentityContextProviderProps } from '@reactionable/amplify';
import { TestWrapper } from '@reactionable/ui-bootstrap';
import { render } from '@testing-library/react';
import React from 'react';
import { MemoryRouter } from 'react-router';
import App from './App';

describe('App', () => {
    
    let identity: IIdentityContextProviderProps;

    beforeAll(() => {
        identity = useIdentityContextProviderProps();
    });

    it('renders without crashing', () => {  
        render(
            <TestWrapper identity={identity}>
                <App />
            </TestWrapper>
        );
    });
});"
`;

exports[`createComponent run should create expected components files 1`] = `
"import React, { FC } from 'react';
import { Helmet } from 'react-helmet';
import { useTranslation } from 'react-i18next';

interface ITestComponentProps { }
const TestComponent: FC<ITestComponentProps> = (props) => {
  const { t } = useTranslation();
  
  return <>
    <Helmet>
      <title>{t('Tests')} - {t('Test component')}</title>
      <meta name=\\"description\\" content={t('Test component description')} />
    </Helmet>
    <h1>{t('Test component')}</h1>
    
  </>;
};
export default TestComponent;"
`;

exports[`createComponent run should create expected components files 2`] = `
"import { IIdentityContextProviderProps, useIdentityContextProviderProps } from '@reactionable/amplify';
import { TestWrapper } from '@reactionable/ui-bootstrap';
import { render } from '@testing-library/react';
import React from 'react';
import { MemoryRouter } from 'react-router';
import TestComponent from './TestComponent';

describe('TestComponent', () => {
    
    let identity: IIdentityContextProviderProps;

    beforeAll(() => {
        identity = useIdentityContextProviderProps();
    });

    it('renders without crashing', () => {  
        render(
            <TestWrapper identity={identity}>
                <TestComponent />
            </TestWrapper>
        );
    });
});"
`;
